src/main/java/com/hse/knopkabackend/services/KnopkaUserService.java

```diff
@@ -0,0 +1,117 @@
+
+    @Autowired
+    public KnopkaUserService(KnopkaUserRepository knopkaUserRepository, ProfileRepository profileKnopkaUserRepository) {
+        this.knopkaUserRepository = knopkaUserRepository;
+        this.profileKnopkaUserRepository = profileKnopkaUserRepository;
+    }
+
+    public List<KnopkaUser> getKnopkaUsers() {
+        return knopkaUserRepository.findAll();
+    }
+
+    public void addNewKnopkaUser(KnopkaUser knopkaUser) {
+        Optional<Profile> profileKnopkaUserByNickname = profileKnopkaUserRepository.findProfileByNickname(
+                knopkaUser.getProfile().getNickname()
+        );
+        if (profileKnopkaUserByNickname.isPresent()) {
+            throw new IllegalStateException("Oops! Nickname '" +
+                    knopkaUser.getProfile().getNickname() + "' is already taken. Try another one."
+            );
+        }
+        Optional<KnopkaUser> knopkaUserByEmail = knopkaUserRepository.findKnopkaUserByEmail(
+                knopkaUser.getEmail()
+        );
+        if (knopkaUserByEmail.isPresent()) {
+            throw new IllegalStateException("Oops! Email '" +
+                    knopkaUser.getEmail() + "' is already taken. Try another one."
+            );
+        }
+        knopkaUserRepository.save(knopkaUser);
+        System.out.println("Added user with unique nickname: " + knopkaUser);
+    }
+
+
+    public void deleteKnopkaUser(Long knopkaUserId, String token) {
+        boolean exists = knopkaUserRepository.existsById(knopkaUserId);
+        if (!exists) {
+            throw new IllegalStateException("KnopkaUser with id: " + knopkaUserId + " doesn't exist");
+        }
+        Optional<KnopkaUser> knopkaUserById = knopkaUserRepository.findById(knopkaUserId);
+        if (knopkaUserById.isPresent() && Objects.equals(token, knopkaUserById.get().getToken())) {
+            knopkaUserRepository.deleteById(knopkaUserId);
+            System.out.println("Deleted KnopkaUser with id: " + knopkaUserId);
+        } else {
+            throw new IllegalStateException("Token is not valid");
+        }
+    }
+
+    @Transactional
+    public void updateKnopkaUserNickname(Long knopkaUserId, String nickname, String token) {
+        KnopkaUser knopkaUser = knopkaUserRepository.findById(knopkaUserId).orElseThrow(
+                () -> new IllegalStateException("KnopkaUser with id: " + knopkaUserId + " doesn't exist")
+        );
+        if (nickname != null) {
+            Optional<Profile> knopkaUserByNickname = profileKnopkaUserRepository.findProfileByNickname(nickname);
+            if (knopkaUserByNickname.isPresent()) {
+                throw new IllegalStateException("Oops! Nickname '" +
+                        nickname + "' is already taken. Try another one."
+                );
+            }
+            if (!nickname.isBlank() && !knopkaUser.getProfile().getNickname().equals(nickname) && Objects.equals(knopkaUser.getToken(), token)) {
```

Comment:
In general, all validation should be moved to a separate method  and then used like
validate(...);   //  throws exception if something is invalid - different exception for different reasons
user.setFiled(newValue);
...
save(user)

Reply:
And there is also validation by annotations (e.g. @NotBlank) and custom validators, just so you know.